src/uu/pgrep/src/pgrep.rs

```diff
@@ -0,0 +1,278 @@
+                .collect()
+        } else if matches.get_flag("list-name") {
+            result
+                .into_iter()
+                .map(|it| format!("{} {}", it.pid, it.clone().status().get("Name").unwrap()))
+                .collect()
+        } else {
+            result.into_iter().map(|it| format!("{}", it.pid)).collect()
+        };
+
+        formatted.join(delimiter)
+    };
+
+    let result = result();
+    if !result.is_empty() {
+        println!("{}", result);
+    };
+
+    Ok(())
+}
+
+fn collect_arg_patterns(matches: &ArgMatches) -> Vec<String> {
+    let should_ignore_case = matches.get_flag("ignore-case");
+
+    let programs = matches
+        .get_many::<String>("pattern")
+        .unwrap_or_default()
+        .map(|it| it.to_string());
+
+    if should_ignore_case {
+        programs.map(|it| it.to_lowercase()).collect::<Vec<_>>()
+    } else {
+        programs.collect()
+    }
+}
+
+fn collect_pid(matches: &ArgMatches, patterns: &[String]) -> Vec<PidEntry> {
+    let should_inverse = matches.get_flag("inverse");
+    let should_ignore_case = matches.get_flag("ignore-case");
+
+    let flag_full = matches.get_flag("full");
+    let flag_exact = matches.get_flag("exact");
+
+    let predicate_closure = |mut it: PidEntry| {
+        let binding = it.status();
+        let name = binding.get("Name")?;
+
+        // Processs flag `--ignore-case`
+        let name = if should_ignore_case {
+            name.to_lowercase()
+        } else {
+            name.into()
+        };
+
+        let mut iter = patterns.iter();
+
+        let name_matched = if flag_full {
+            // Equals `Name` in /proc/<pid>/status
+            iter.any(|it| name.eq(it))
+        } else if flag_exact {
+            // Equals `cmdline` in /proc/<pid>/cmdline
+            iter.any(|s| it.cmdline.eq(s))
+        } else {
+            // Contains `cmdline` in /proc/<pid>/cmdline
+            iter.any(|it| name.contains(it))
+        };
+
+        let tty_matched = if let Some(ttys) = matches.get_many::<String>("terminal") {
+            // convert from input like `pts/0`
+            let ttys = ttys
+                .cloned()
+                .flat_map(TerminalType::try_from)
+                .collect::<HashSet<_>>();
+
+            if let Ok(value) = it.ttys() {
+                value.iter().any(|it| ttys.contains(it))
+            } else {
+                false
+            }
+        } else {
+            true
+        };
+
+        if (name_matched || tty_matched) ^ should_inverse {
+            Some(it)
+        } else {
+            None
+        }
+    };
+
+    let mut result = Vec::new();
+
+    for ele in walk_pid() {
+        if let Some(ele) = predicate_closure(ele) {
+            result.push(ele)
+        }
+    }
+
+    result
+}
+
+// Make -o and -n as a group of args
+fn collect_oldest_newest(matches: &ArgMatches, patterns: &[String]) -> Option<Vec<PidEntry>> {
```

Comment:
Why do you return an `Option<Vec<PidEntry>>`? Conceptually, there can be at most one newest/oldest pid and so `Option<PidEntry>` should be enough.

Reply:
I want `handle_oldest_newest` and `handle_matching_pids` to have the same return type because they are mutually exclusive and one and only one of the functions will execute properly. This makes it easy to unify the handling of detail flags like `-l` `-a` and so on.

Reply:
Hm, to me it looks like you are solving something at the code level that could be solved at the conceptual level.

Forget the code for a moment and think about what (a part of) `pgrep` on a conceptual level does:

* it creates a list of PIDs,
* sorts the list,
* filters the list,
* and uses the list to generate some output

"some output" can be whatever: all elements of the list, the first element, the last element, the number of elements.

Now, back to the code. Based on their name, I would put `handle_matching_pids` on the third step of this simple model whereas `handle_oldest_newest` is on the fourth step. And so there is a conflict on the conceptual level if you look at those functions as mutually exclusive ;-)

Of course, it's possible that this model is incorrect/incomplete, but I think it provides at least some food for thought.

Reply:
Oh, that's exactly what I was thinking! If I change the `handle_*` function to be a filter, then the whole code will be easier to understand